src/hotspot/share/runtime/objectMonitor.cpp

```diff
@@ -475,6 +492,110 @@ int ObjectMonitor::TryLock(Thread * Self) {
   return -1;
 }
 
+// Deflate the specified ObjectMonitor if not in-use. Returns true if it
+// was deflated and false otherwise.
+//
+// The async deflation protocol sets owner to DEFLATER_MARKER and
+// makes contentions negative as signals to contending threads that
+// an async deflation is in progress. There are a number of checks
+// as part of the protocol to make sure that the calling thread has
+// not lost the race to a contending thread.
+//
+// The ObjectMonitor has been successfully async deflated when:
+//   (contentions < 0)
+// Contending threads that see that condition know to retry their operation.
+//
+bool ObjectMonitor::deflate_monitor() {
+  if (is_busy()) {
+    // Easy checks are first - the ObjectMonitor is busy so no deflation.
+    return false;
+  }
+
+  if (ObjectSynchronizer::is_final_audit() && owner_is_DEFLATER_MARKER()) {
+    // The final audit can see an already deflated ObjectMonitor on the
+    // in-use list because MonitorList::unlink_deflated() might have
+    // blocked for the final safepoint before unlinking all the deflated
+    // monitors.
+    assert(contentions() < 0, "must be negative: contentions=%d", contentions());
+    // Already returned 'true' when it was originally deflated.
+    return false;
+  }
+
+  const oop obj = object_peek();
+
+  if (obj == NULL) {
+    // If the object died, we can recycle the monitor without racing with
+    // Java threads. The GC already broke the association with the object.
+    set_owner_from(NULL, DEFLATER_MARKER);
+    assert(contentions() >= 0, "must be non-negative: contentions=%d", contentions());
+    _contentions = -max_jint;
+  } else {
+    // Attempt async deflation protocol.
+
+    // Set a NULL owner to DEFLATER_MARKER to force any contending thread
+    // through the slow path. This is just the first part of the async
+    // deflation dance.
+    if (try_set_owner_from(NULL, DEFLATER_MARKER) != NULL) {
+      // The owner field is no longer NULL so we lost the race since the
+      // ObjectMonitor is now busy.
+      return false;
+    }
+
+    if (contentions() > 0 || _waiters != 0) {
+      // Another thread has raced to enter the ObjectMonitor after
+      // is_busy() above or has already entered and waited on
+      // it which makes it busy so no deflation. Restore owner to
+      // NULL if it is still DEFLATER_MARKER.
+      if (try_set_owner_from(DEFLATER_MARKER, NULL) != DEFLATER_MARKER) {
+        // Deferred decrement for the JT EnterI() that cancelled the async deflation.
+        add_to_contentions(-1);
```

Comment:
No it is not a ref_count. We got rid of the accurate ref_count field
because maintaining it was too slow.

contentions just tells you how many threads have blocked on the
slow-path trying to enter the monitor. The fast-paths never touch
contentions. The comment on line 360 is accurate.

Reply:
But contentions is used for more than informing JVMTI, it's used to test whether the monitor is_busy on the slow path.  That's why I wanted the comment to say something like your last sentence, since I spent time trying to understand why the various calls to add_to_contentions(-1) in deflate_monitor earlier.

Reply:
Ahhh... I think I understand your confusion. This line:

L550: // Deferred decrement for the JT EnterI() that cancelled the async deflation.
L551:  add_to_contentions(-1);

doesn't match up with this line:

L361: add_to_contentions(1);

It matches up with one of these:
```
  if (try_set_owner_from(DEFLATER_MARKER, Self) == DEFLATER_MARKER) {
    // Cancelled the in-progress async deflation by changing owner from
    // DEFLATER_MARKER to Self. As part of the contended enter protocol,
    // contentions was incremented to a positive value before EnterI()
    // was called and that prevents the deflater thread from winning the
    // last part of the 2-part async deflation protocol. After EnterI()
    // returns to enter(), contentions is decremented because the caller
    // now owns the monitor. We bump contentions an extra time here to
    // prevent the deflater thread from winning the last part of the
    // 2-part async deflation protocol after the regular decrement
    // occurs in enter(). The deflater thread will decrement contentions
    // after it recognizes that the async deflation was cancelled.
    add_to_contentions(1);
```

The long comments are in the two places where we temporarily
increment contentions to stop the race with the deflater thread
and the shorter comment, e.g., L550, are for where we undo the
temporary increment.

The primary purpose of the contentions field is for JVM/TI and
M&M queries. We just (temporarily) steal it for async deflation
purposes...

Reply:
Well since it controls async deflation, it should probably get a mention since this comment on its own is not true:

// Keep track of contention for JVM/TI and M&M queries and control async deflation.

Reply:
The comment for **that** location is correct. That's the location
where we keep track of contended monitors for JVM/TI and M&M.
The comments at the other locations are also correct (very verbose
for two of them, but correct).

Reply:
Look at the entire context of that location: